File: backend/app/schemas/ocr.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime


class OCRActionRequest(BaseModel):
    action: str  # "summary" | "explain" | "mcqs" | "flashcards" | "translate" | "questions"
    extracted_text: str
    target_language: Optional[str] = "Tamil"  # for translation


class OCRScanResponse(BaseModel):
    id: int
    user_id: int
    image_name: str
    image_url: Optional[str] = None
    extracted_text: str
    language_detected: str
    confidence_score: float
    summary: Optional[str] = None
    flashcards: Optional[str] = None
    mcqs: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class OCRScanListResponse(BaseModel):
    id: int
    user_id: int
    image_name: str
    language_detected: str
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/api/ocr.ts
import { fetchApi, API_URL } from '../lib/api';

export interface OCRScanResponse {
  id: number;
  user_id: number;
  image_name: string;
  image_url?: string;
  extracted_text: string;
  language_detected: string;
  confidence_score: number;
  summary?: string;
  flashcards?: string;
  mcqs?: string;
  created_at: string;
}

export interface OCRScanListItem {
  id: number;
  user_id: number;
  image_name: string;
  language_detected: string;
  created_at: string;
}

export interface OCRActionPayload {
  action: 'summary' | 'explain' | 'mcqs' | 'flashcards' | 'translate' | 'questions';
  extracted_text: string;
  target_language?: string;
}

export const uploadAndExtractOCR = async (file: File): Promise<OCRScanResponse> => {
  const formData = new FormData();
  formData.append('file', file);

  const token = localStorage.getItem('token');
  const headers: Record<string, string> = {};
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  const response = await fetch(`${API_URL}/ocr/upload`, {
    method: 'POST',
    headers,
    body: formData,
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Failed to extract text from image');
  }

  return await response.json();
};

export const executeAIAction = async (payload: OCRActionPayload): Promise<{ action: string; result: string }> => {
  return await fetchApi('/ocr/action', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
};

export const getOCRHistory = async (): Promise<OCRScanListItem[]> => {
  return await fetchApi('/ocr/history');
};

export const getOCRScanById = async (id: number): Promise<OCRScanResponse> => {
  return await fetchApi(`/ocr/${id}`);
};

export const deleteOCRScan = async (id: number): Promise<void> => {
  await fetchApi(`/ocr/${id}`, { method: 'DELETE' });
};
